### PR TITLE
#1 Restrict setup call to same-origin

### PR DESCRIPTION
**Pull request by ferjm**
r? @michielbdejong

### CODE REVIEW

File: src/users_router.rs [hunk 1]
```diff
@@ -15,8 +15,37 @@ use self::unicase::UniCase;
 struct CORS;
 
 impl AfterMiddleware for CORS {
-    fn after(&self, _: &mut Request, mut res: Response)
+    fn after(&self, req: &mut Request, mut res: Response)
         -> IronResult<Response> {
+
+        // Endpoints listed here will be denied CORS.
+        // Endpoints containing a variable path can use '*' like in:
+        // &["users", "*"]
```
Comment (michielbdejong) [inline src/users_router.rs]: Isn't that equivalent to `&["*"]`?

File: src/users_router.rs [hunk 2]
```diff
@@ -15,8 +15,37 @@ use self::unicase::UniCase;
...
+        const SAME_ORIGIN_ONLY_ENDPOINTS: &'static[&'static[&'static str]] =
+            &[&["setup"]];
+
+        let mut is_soo_endpoint = false;
+        for endpoint in SAME_ORIGIN_ONLY_ENDPOINTS {
+            if endpoint.len() != req.url.path.len() {
+                continue;
```
Comment (michielbdejong) [inline src/users_router.rs:50]: Doesn't that make `"*"` only apply for URL paths of length 1?

**Comment by michielbdejong** [comment]
It would feel safer to me to list the cross-origin endpoints, and default to single-origin.

It's probably worth adding some unit tests.

File: src/users_router.rs [hunk 3]
```diff
@@ -7,16 +7,68 @@ extern crate iron;
...
         -> IronResult<Response> {
+
+        let mut is_cors_endpoint = false;
+        for endpoint in CORS::ENDPOINTS {
+            let (ref method, path) = *endpoint;
+            if req.method != *method {
+                continue;
+            }
+            if path.len() != req.url.path.len() {
+                continue;
```
Comment (michielbdejong) [inline src/users_router.rs:53]: Same question - https://github.com/fxbox/users/pull/4#discussion_r52615773

File: src/users_router.rs [hunk 4]
```diff
@@ -7,16 +7,68 @@ extern crate iron;
...
+                    break;
+                }
+                is_cors_endpoint = true;
+            }
+            if is_cors_endpoint {
+                break;
+            }
+        }
+
+        if !is_cors_endpoint {
```
Comment (michielbdejong) [inline src/users_router.rs:68]: Nit: you can do `if is_cors_endpoint {` and then put lines 72 - 77 inside the if-statement.

File: src/users_router.rs [hunk 5]
```diff
@@ -64,3 +116,54 @@ impl UsersRouter {
...
+            body: unsafe { ::std::mem::uninitialized() },
+            method: method.clone(),
+            extensions: TypeMap::new()
+        }
+    }
+
+    // Test that all CORS allowed endpoints get the appropriate CORS headers.
+    for endpoint in CORS::ENDPOINTS {
+        let (ref method, path) = *endpoint;
+        let path = path.join("/");
```
Comment (michielbdejong) [inline src/users_router.rs]: This becomes e.g. `permissions/_/*` and not something like `permissions/_/foo`, ~~which is probably the reason this test is passing :)~~

You could replace all occurrences of the string `*` with the string `foo` to get a more realistic test.

**Comment by michielbdejong** [comment]
r+wc: please fix the `*` case in ~~both the implementation and~~ the test.

File: src/users_router.rs (hunk 3 above)
Comment (michielbdejong) [inline src/users_router.rs:53]: Ah right, `path` and `req.url.path` are arrays, sorry. Please ignore. :)

**Comment by ferjm** [comment]
Thanks!

**Pull request merged**